### ISSUE TITLE
Method return type not being inferred for templated class

### ISSUE THREAD

**Issue by joshdifabio** [vimeo/psalm#1600]
I don't know how best to describe this issue, but I think the repro will make it fairly clear:

https://psalm.dev/r/c8661b5bb2

Psalm is not inferring that `$optional->get()` returns `string`, even though it has correctly inferred that `$optional` is `Optional<string>`.

**Actual message**
```
INFO: MissingReturnType - 29:10 - Method expandOptions does not have a return type, expecting Collection<mixed>
```

**Expected message**
```
INFO: MissingReturnType - 29:10 - Method expandOptions does not have a return type, expecting Collection<string>
```

**Comment by weirdan** [comment]
Psalm does not infer closure type from context, and it needs its type to infer the resulting type of `map()`.
https://psalm.dev/r/b6908489aa is the way to make it pass.

**Comment by muglug** [comment]
Yeah, it only checks the boundaries (so it’d complain if the param was `Optional<int>`). There’s some special handling, IIRC, for `array_map`, but that’s about it.

**Comment by joshdifabio** [comment]
That's a shame. Is that a technical limitation or a design choice? I'd rather not have to write things twice if they can be inferred. I'd rather not have any type declaration on my closures at all if Psalm can infer the types, especially once arrow functions land.

**Comment by joshdifabio** [comment]
@muglug I'm not sure if GitHub notifies for comments in closed issues, so ... ^

**Comment by muglug** [comment]
Oh it does notify.

Generally things are just much simpler when you treat closures as boundaries for type inference. That way you can analyse the closure body, verify the function works as expected, and separately verify that the stated closure signature matches the one that the closure's being passed to.

Consider this code:

```php
/**
 * @param Collection<Optional<string>> $collection
 */
function expandOptions(Collection $collection) {
  $a = function (Optional $optional) {
    return $optional->get();
  };
  return $collection->map($a);
}
```

We shouldn't have to re-analyse the closure body `$a` after it's passed to `map` - analysing it once should be sufficient.

Another potential issue here is that Psalm doesn't _insist_ on a signature param type having all its signature params defined (so `Optional` is allowed where it should probably insist on a `Optional<...>` annotation). Feel free to ticket that.

**Comment by muglug** [comment]
Psalm's behaviour is also different from HHVM's - I've opened a ticket at https://github.com/facebook/hhvm/issues/8496 because maybe I'm just wrong.

**Comment by joshdifabio** [comment]
> We shouldn't have to re-analyse the closure body $a after it's passed to map - analysing it once should be sufficient.

You might be right, one could argue either way, but that's a bit different to my case because you're assigning the closure to a variable when declaring it, at which point you can't necessarily infer param and return types, whereas I'm declaring a closure in a context within which the types could be inferred. The two cases are treated differently by TypeScript and Java I believe, and I would assume most other implementations of generics.

To clarify:

```php
/**
 * @param Collection<Optional<string>> $collection
 */
function expandOptions(Collection $collection) {
  // we can't infer the param/return types yet -- fair enough
  $a = function (Optional $optional) {
    return $optional->get();
  };
  return $collection->map($a);
}
```

versus

```
/**
 * @param Collection<Optional<string>> $collection
 */
function expandOptions(Collection $collection) {
  // we can safely infer the param and return types here
  return $collection->map(function ($optional) {
    return $optional->get();
  });
}
```

I can certainly understand the first approach requiring a hint above the assignment, but for the very common case of passing a closure as an argument to another function, which will become even more common with arrow functions, it just adds noise, which is why in most other languages those types are inferred and developers don't specify them.

I'm afraid it's not clear to me whether the lack of inference in this case is your preferred approach regardless of implementation complexity within Psalm itself, or if it's your preference because of the complexity it would add to Psalm to infer param and return types in these cases. If it's the latter then it's probably not worth us discussing any further -- you're the expert, and if it's too hard to do then it's too hard to do!

**Comment by muglug** [comment]
> The two cases are treated differently by TypeScript and Java I believe

Can you provide examples?

**Comment by joshdifabio** [comment]
Here is a TypeScript example. You can ignore the transpiled JavaScript.

https://www.typescriptlang.org/play/index.html#src=interface%20Collection%3CE%3E%20%7B%0D%0A%20%20%20%20map%3CR%3E(mapper%3A%20(element%3A%20E)%20%3D%3E%20R)%3A%20Collection%3CR%3E%3B%0D%0A%7D%0D%0A%0D%0Ainterface%20Optional%3CT%3E%20%7B%0D%0A%20%20%20%20get()%3A%20T%3B%0D%0A%7D%0D%0A%0D%0Afunction%20expandOptions1%3CT%3E(optionals%3A%20Collection%3COptional%3CT%3E%3E)%3A%20Collection%3CT%3E%20%7B%0D%0A%20%20%20%20return%20optionals.map(e%20%3D%3E%20e.foo())%3B%0D%0A%7D%0D%0A%0D%0Afunction%20expandOptions2%3CT%3E(optionals%3A%20Collection%3COptional%3CT%3E%3E)%3A%20Collection%3CT%3E%20%7B%0D%0A%20%20%20%20const%20fn%20%3D%20e%20%3D%3E%20e.foo()%3B%0D%0A%20%20%20%20return%20optionals.map(fn)%3B%0D%0A%7D

**Comment by muglug** [comment]
Yeah, you've convinced me that this is something Psalm should do.

**Comment by muglug** [comment]
I'm going to solve it for when param types aren't given, not sure about return types.

**Comment by muglug** [comment]
This now works: https://psalm.dev/r/df23fd1a67

**Comment by muglug** [comment]
And it'll be even better [in PHP 7.4](https://wiki.php.net/rfc/arrow_functions_v2) - 

```php
return $collection->map(function ($o) { return $o->get(); });
```

will become

```php
return $collection->map(fn($o) => $o->get());
```

**Comment by weirdan** [comment]
That works with single-closure expression only, right? `$collection->map($f = function($o) { return $o->get(); });` is out of scope?

**Comment by muglug** [comment]
Yeah, that won't work

**Comment by joshdifabio** [comment]
This is fantastic -- thank you!
